compose/utils.py

```diff
@@ -17,58 +17,50 @@
 json_decoder = json.JSONDecoder()
 
 
-def parallel_execute(objects, obj_callable, msg_index, msg):
-    """
-    For a given list of objects, call the callable passing in the first
+def perform_operation(func, arg, callback, index):
+    try:
+        callback((index, func(arg)))
+    except Exception as e:
+        callback((index, e))
+
+
+def parallel_execute(objects, func, index_func, msg):
+    """For a given list of objects, call the callable passing in the first
     object we give it.
     """
     stream = get_output_stream(sys.stdout)
-    lines = []
+    writer = ParallelStreamWriter(stream, msg)
 
     for obj in objects:
-        write_out_msg(stream, lines, msg_index(obj), msg)
+        writer.initialize(index_func(obj))
 
     q = Queue()
 
-    def inner_execute_function(an_callable, parameter, msg_index):
-        error = None
-        try:
-            result = an_callable(parameter)
-        except APIError as e:
-            error = e.explanation
-            result = "error"
-        except Exception as e:
-            error = e
-            result = 'unexpected_exception'
-
-        q.put((msg_index, result, error))
-
-    for an_object in objects:
+    # TODO: limit the number of threads #1828
+    for obj in objects:
         t = Thread(
-            target=inner_execute_function,
-            args=(obj_callable, an_object, msg_index(an_object)),
-        )
+            target=perform_operation,
+            args=(func, obj, q.put, index_func(obj)))
         t.daemon = True
         t.start()
 
     done = 0
     errors = {}
-    total_to_execute = len(objects)
 
-    while done < total_to_execute:
+    while done < len(objects):
         try:
-            msg_index, result, error = q.get(timeout=1)
-
-            if result == 'unexpected_exception':
-                errors[msg_index] = result, error
-            if result == 'error':
-                errors[msg_index] = result, error
-                write_out_msg(stream, lines, msg_index, msg, status='error')
-            else:
-                write_out_msg(stream, lines, msg_index, msg)
-            done += 1
+            msg_index, result = q.get(timeout=1)
         except Empty:
-            pass
+            continue
+
+        if isinstance(result, APIError):
+            errors[msg_index] = "error", result
+            writer.write(msg_index, 'error')
+        elif isinstance(result, Exception):
+            errors[msg_index] = "unexpected_exception", result
+        else:
+            writer.write(msg_index, 'done')
```

Comment:
I noticed we had effectively the same branch conditions in both the nested function and the "collect" phase. Instead of having both, we can just check the result type for errors.